django/tasks/task.py

```diff
@@ -0,0 +1,271 @@
+        self.get_backend().validate_task(self)
+
+    @property
+    def name(self):
+        """
+        An identifier for the task
+        """
+        return self.func.__name__
+
+    def using(
+        self,
+        *,
+        priority=None,
+        queue_name=None,
+        run_after=None,
+        backend=None,
+    ):
+        """
+        Create a new task with modified defaults
+        """
+
+        changes = {}
+
+        if priority is not None:
+            changes["priority"] = priority
+        if queue_name is not None:
+            changes["queue_name"] = queue_name
+        if run_after is not None:
+            changes["run_after"] = run_after
+        if backend is not None:
+            changes["backend"] = backend
+
+        return replace(self, **changes)
+
+    def enqueue(self, *args, **kwargs):
+        """
+        Queue up the task to be executed
+        """
+        return self.get_backend().enqueue(
+            self, json_normalize(args), json_normalize(kwargs)
+        )
+
+    async def aenqueue(self, *args, **kwargs):
+        """
+        Queue up a task function (or coroutine) to be executed
+        """
+        return await self.get_backend().aenqueue(
+            self, json_normalize(args), json_normalize(kwargs)
+        )
+
+    def get_result(self, result_id):
+        """
+        Retrieve the result for a task of this type by its id (if one exists).
+        If one doesn't, or is the wrong type, raises ResultDoesNotExist.
+        """
+        result = self.get_backend().get_result(result_id)
+
+        if result.task.func != self.func:
+            raise ResultDoesNotExist
+
+        return result
+
+    async def aget_result(self, result_id):
+        """
+        Retrieve the result for a task of this type by its id (if one exists).
+        If one doesn't, or is the wrong type, raises ResultDoesNotExist.
+        """
+        result = await self.get_backend().aget_result(result_id)
+
+        if result.task.func != self.func:
+            raise ResultDoesNotExist
+
+        return result
+
+    def call(self, *args, **kwargs):
+        if iscoroutinefunction(self.func):
+            return async_to_sync(self.func)(*args, **kwargs)
+        return self.func(*args, **kwargs)
+
+    async def acall(self, *args, **kwargs):
+        if iscoroutinefunction(self.func):
+            return await self.func(*args, **kwargs)
+        return await sync_to_async(self.func)(*args, **kwargs)
+
+    def get_backend(self):
+        from . import tasks
+
+        return tasks[self.backend]
+
+    @property
+    def module_path(self):
+        return get_module_path(self.func)
+
+
+# Implementation
+def task(
+    function=None,
+    *,
+    priority=DEFAULT_PRIORITY,
+    queue_name=DEFAULT_QUEUE_NAME,
+    backend=DEFAULT_TASK_BACKEND_ALIAS,
+    enqueue_on_commit=None,
+):
+    """
+    A decorator used to create a task.
+    """
+    from . import tasks
+
+    def wrapper(f):
```

Comment:
Add `@functools.wraps(function)` to preserve docstrings and such?

Reply:
This wrapper itself is very short lived. It's either called immediately inside this function, or immediately once scope return to call the parent. I don't think it's necessary to add that too.

`functools.wraps` updates a fair bit about the function (eg type annotations) which also isn't quite right here.